lib/lveditqmljs/src/codeqmlhandler.cpp

```diff
@@ -3819,34 +3798,42 @@ bool CodeQmlHandler::isForAnObject(const lv::QmlDeclaration::Ptr &declaration){
     return QmlTypeInfo::isObject(declaration->type().name());
 }
 
-PaletteList *CodeQmlHandler::findPalettesForDeclaration(QmlDeclaration::Ptr declaration, bool includeLayoutConfigurations)
+QJSValue CodeQmlHandler::findPalettesForDeclaration(QmlDeclaration::Ptr declaration)
 {
     // every fragment has a declaration -> should end up here
 
     Q_D(CodeQmlHandler);
 
-    PaletteContainer::PaletteSearch configurations = includeLayoutConfigurations ? PaletteContainer::IncludeLayoutConfigurations : PaletteContainer::Empty;
-
-    PaletteList* lpl = d->projectHandler->paletteContainer()->findPalettes(declaration->type().join(), configurations);
+    QList<PaletteLoader*> result = d->projectHandler->paletteContainer()->findPalettes(declaration->type().join());
 
     if (declaration->isForComponent()){
-        lpl = d->projectHandler->paletteContainer()->findPalettes("qml/Component", configurations, lpl);
+        auto components = d->projectHandler->paletteContainer()->findPalettes("qml/Component");
+        result.append(components);
     } else {
-        if ( declaration->isForObject())
-        if (declaration->type().name()[0].isUpper() && declaration->type().language() == QmlTypeReference::Qml){
-            lpl = d->projectHandler->paletteContainer()->findPalettes("qml/Object", configurations, lpl);
+        if ( declaration->isForObject()){
+            if (declaration->type().name()[0].isUpper() && declaration->type().language() == QmlTypeReference::Qml){
+                auto object = d->projectHandler->paletteContainer()->findPalettes("qml/Object");
+                result.append(object);
+            }
         }
 
         if ( declaration->isForList() ){
-            lpl = d->projectHandler->paletteContainer()->findPalettes("qml/childlist", configurations, lpl);
+            auto child = d->projectHandler->paletteContainer()->findPalettes("qml/childlist");
+            result.append(child);
         } else {
-            lpl = d->projectHandler->paletteContainer()->findPalettes("qml/property", configurations, lpl);
+            auto prop = d->projectHandler->paletteContainer()->findPalettes("qml/property");
+            result.append(prop);
         }
     }
 
+    QJSValue res = m_engine->engine()->newArray(result.length() + 1);
+    for (int i = 0; i < result.length(); ++i){
+        res.setProperty(i, d->projectHandler->paletteContainer()->paletteData(result[i]));
+    }
+
+    res.setProperty(result.length(), QJSValue(declaration->position() + (declaration->isForImports() ? 7 : 0)));
```

Comment:
It's quite unintuitive to have the position as the last index of the list. Could we return the entire declaration then?

```js
return {
   data: [...]. // list of palettes
   declaration: {. // the entire declaration info object
      position: 0,
      length: 0,
      type: 'qml/imports'
      ...
   }
}
```

We could:

 * create a private function in CodeQmlHandler `QJSValue declarationToQml(QmlDeclaration::Ptr)`
 * copy the implementation from `CodeQMLhandler::declarationInfo`
 * call the function from `CodeQmlHandler::declarationInfo` as well:

```cpp
    if ( !m_document )
        return QJSValue();

    QList<QmlDeclaration::Ptr> properties = getDeclarations(position, length);
    if ( properties.isEmpty() )
        return QJSValue();

    auto declaration = properties.first();
    return declarationToQml(declaration)
```

 * rewrite the result property from `findPalettesForDeclaration` to be an object containing the `data` property, which is the list of palettes, and the `declaration` property which is the declaration info.